src/compartment.c

```diff
@@ -0,0 +1,526 @@
+{
+    Elf64_Auxinfo new_auxv = {at_type, {at_val} };
+    comp_stack_push(comp, &new_auxv, sizeof(new_auxv));
+}
+
+void
+comp_map(struct Compartment* to_map)
+{
+    assert(!(to_map->mapped || to_map->mapped_full));
+    struct SegmentMap* curr_seg;
+    void* map_result;
+
+    for (size_t i = 0; i < to_map->seg_count; ++i)
+    {
+        curr_seg = to_map->segs[i];
+        printf("Currently mapping segment %lu:\n", i);
+        segmap_print(curr_seg);
+        if (curr_seg->file_sz == curr_seg->mem_sz)
+        {
+            map_result = mmap((void*) curr_seg->mem_bot,
+                                    curr_seg->file_sz,
+                                    /*curr_seg->prot_flags,*/ // TODO currently need read/write to inject the intercepts, consider better option
+                                    PROT_READ | PROT_WRITE | PROT_EXEC,
+                                    MAP_PRIVATE | MAP_FIXED,
+                                    to_map->fd, curr_seg->offset);
+        }
+        else
+        {
+            assert(curr_seg->mem_sz > curr_seg->file_sz);
+            map_result = mmap((void*) curr_seg->mem_bot,
+                                    curr_seg->mem_sz,
+                                    curr_seg->prot_flags,
+                                    MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+                                    -1, 0);
+            assert(map_result !=  MAP_FAILED);
+            int pread_res = pread(to_map->fd, (void*) curr_seg->mem_bot,
+                                  curr_seg->file_sz, curr_seg->offset);
+            assert(pread_res != -1);
+        }
+    }
+
+    printf("Currently mapping compartment scratch memory + stack\n");
+    map_result = mmap((void*) to_map->scratch_mem_base,
+                      to_map->scratch_mem_size,
+                      PROT_READ | PROT_WRITE,
+                      MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+                      -1, 0);
+    assert(map_result != MAP_FAILED);
+    map_result = mmap((void*) to_map->scratch_mem_stack_top - to_map->scratch_mem_stack_size,
+                      to_map->scratch_mem_stack_size,
+                      PROT_READ | PROT_WRITE,
+                      MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS | MAP_STACK,
+                      -1, 0);
+    to_map->stack_pointer = to_map->scratch_mem_stack_top;
+    assert(map_result != MAP_FAILED);
+
+    for (size_t i = 0; i < to_map->relas_cnt; ++i)
+    {
+        uintptr_t rela_addr = to_map->relas[i] + to_map->base;
+        uintptr_t old_plt_val = (uintptr_t) *((void**) rela_addr);
+        old_plt_val += to_map->base;
+        *((uintptr_t *) rela_addr) = old_plt_val;
+    }
+
+    for (size_t i = 0; i < to_map->curr_intercept_count; ++i)
+    {
+        struct intercept_patch to_patch = to_map->patches[i];
+        *to_patch.patch_addr = to_patch.instr;
+    }
+
+    to_map->mapped = true;
+}
+
+void
+comp_map_full(struct Compartment* to_map)
```

Comment:
I'm not quite sure I understand this function (maybe because I don't quite understand the name).

Reply:
This was something I used to just bulk map all was needed for an ELF file, rather than carefully map each segment. Used it at some point to try things. I've removed it in https://github.com/capablevms/CHERI-ELF-comp/pull/1/commits/c917047d111241c170fdb850cd136619ab8d774b, since it shouldn't be needed anymore.